### PR TITLE
Complete nullable annotations

### PR DESCRIPTION
**Pull request by iamcarbon**
This PR completes the nullability annotations and enables nullable at the project level. 

There are 4 possibility breaking changes introduced to enforce initialization of non-nullable properties to the following classes.

HeadingLinkReferenceDefinition
FootnoteLink
FootnoteLinkReference
HtmlInline

While these classes have to be constructed with a non-null object, they remain fully mutable after construction to allow the DOM to be manipulated.

### CODE REVIEW

File: src/Markdig/Helpers/StringSlice.cs [hunk 1]
```diff
@@ -81,7 +83,7 @@ public StringSlice(string text, int start, int end, NewLine newLine)
         /// <summary>
         /// The text of this slice.
         /// </summary>
-        public readonly string Text;
+        public string Text;
```
Comment (MihaZupan) [inline src/Markdig/Helpers/StringSlice.cs:86]: Why this change?

File: src/Markdig/Renderers/Normalize/Inlines/LinkInlineRenderer.cs [hunk 2]
```diff
@@ -44,7 +44,7 @@ protected override void Write(NormalizeRenderer renderer, LinkInline link)
                 {
                     renderer.Write('(').Write(link.Url);
 
-                    if (!string.IsNullOrEmpty(link.Title))
+                    if (link.Title is { Length: > 0 })
```
Comment (MihaZupan) [inline src/Markdig/Renderers/Normalize/Inlines/LinkInlineRenderer.cs:47]: We don't *have* to change everything to pattern matching. Last time I checked it still had worse code generation, so as long as we don't do it on hot paths is should be fine.

**Review by MihaZupan — COMMENTED** [review]
Aside from the StringSlice change, LGTM, thanks

File: src/Markdig/Helpers/StringSlice.cs (hunk 1 above)
Comment (iamcarbon) [reply src/Markdig/Helpers/StringSlice.cs]: Good catch! Thank you. This was accidental when backing out of changing this to a readonly property.

File: src/Markdig/Renderers/Normalize/Inlines/LinkInlineRenderer.cs (hunk 2 above)
Comment (iamcarbon) [reply src/Markdig/Renderers/Normalize/Inlines/LinkInlineRenderer.cs:47]: Agreed -- and have seen similar reports of worse codegen for this pattern. This is another one-off to ensure for the analyzer to flow through the non-null check.

**Comment by iamcarbon** [comment]
@MihaZupan -- Thank you for reviewing and catching that unintended change. Going to squeeze in one final commit to remove the existing #nullable enable directives and reduce churn.... and this should be GTG.

**Comment by iamcarbon** [comment]
And done.

**Comment by xoofx** [comment]
Amazing work @iamcarbon! 👍
